hazelcast/src/main/java/com/hazelcast/jet/impl/execution/init/ExecutionPlan.java

```diff
@@ -377,8 +391,8 @@ public JobConfig getJobConfig() {
         return jobConfig;
     }
 
-    void addVertex(VertexDef vertex) {
-        vertices.add(vertex);
+    synchronized void setVertex(int position, VertexDef vertex) {
```

Comment:
I couldn't completely understand the reason for changing it to a synchronized method. Could you explain with which accesses are we trying to synchronize this setVertex? I am asking it since only making this method synchronized alone may not be enough.

Reply:
Now all vertices' meta suppliers initialization may be done on different thread; see `ExecutionPlanBuilder#createExecutionPlans`, where we interate over each vertex, init it and add to the plan. Because different threads may add vertices, we must synchronize access to the collection, otherwise different threads may overwrite it's writes

Reply:
If we only need to synchronize the writes to different indexes on the collection, this synchronized block may not be necessary after replacing this ArrayList with a fixed sized collection. Of course, the effect of this on performance may be negligible, I'm just writing it to just consider it. cc: @krzysztofslusarski

Reply:
If we can switch to fixed size array then that ```synchronized``` is not needed. 

You also shouldn't use ```synchronized``` on ```this```, better way is to create an ```private static final Object MUX = new Object();``` and synchronize on that. The difference is with ```synchronized``` on ```this``` you have no control if someone do:

```java
ExecutionPlan executionPlan = new ExecutionPlan();
synchronized (executionPlan) {
    // ...
}
```